### PR TITLE
Fix water coil user-specified sizing and Component Sizing Summary table for fan coils

### LINKED ISSUE #7914: Coil:Cooling:Water in Component Sizing Summary has incorrect Design Size and User-Specified Air Flow Rates
Issue overview
--------------
I ran a model using ZoneHVAC:FourPipeFanCoil objects where I hard-sized the "Maximum Supply Air Flow Rate" to 449.15 CFM and I left the Coil:Cooling:Water "Design Air Flow Rate" input set to 'Autosize'.  In the Component Sizing Summary, the ZoneHVAC:FourPipeFanCoil table shows a Design Size Maximum Supply Air Flow Rate of 1097.17 CFM and a User-Specified Maximum Supply Air Flow Rate of 449.15 CFM for one of the zones.  The Coil:Cooling:Water table shows a Design Size Design Air Flow Rate of 449.15 CFM and a User-Specified Design Air Flow Rate of 1097.27 CFM.  These values appear to be switched between the ZoneHVAC:FourPipeFanCoil table and the Coil:Cooling:Water table. 

image

### Details
Some additional details for this issue (if relevant):
 - EnergyPlus v9.3

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #7998: Incorrect Coil:Cooling:Water sizing when hard-sizing ZoneHVAC:FourPipeFanCoil air flow rate
Issue overview
--------------
I ran a model using ZoneHVAC:FourPipeFanCoil objects where I hard-sized the "Maximum Supply Air Flow Rate" to 449.15 CFM and I left the Coil:Cooling:Water "Design Air Flow Rate" input set to 'Autosize'. When running this model through the debugger, I found that the Coil:Cooling:Water was being sized based on the autosized coil design air flow rate of 1097.27 CFM instead of the 449.15 CFM set in the ZoneHVAC:FourPipeFanCoil object.  The same phenomena may also occur with the hot water coil and water flow rates.

Possibly related to https://github.com/NREL/EnergyPlus/issues/7276.

### Details
Some additional details for this issue (if relevant):
 - EnergyPlus v8.3

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by matthew-larson**
Pull request overview
---------------------
 - Fixes #7914 and #7998 
 - This pull request addresses incorrect sizing in the chilled water coils when the water flow and/or air flow are specified for water coils referenced by ZoneHVAC:FourPipeFanCoil objects.  For the pre-sizing data calculations, I modified the DataWaterFlowUsedForSizing to Autosize, similar to the other pre-sizing data parameters.  In order to pass in the original design parameter value for the RequestSizing function, it now references the DataAirFlowUsedForSizing and DataWaterFlowUsedForSizing, which outputs these as the Design Size values in the Component Sizing Summary table and the hard-sized values as User-Specified values.  This also ensures the user-specified values are the values being used for the sizing calculations for coil capacity.
- The unit test is in the WaterCoils.unit.cc but uses a built-up zone fan coil system that verifies the final design air flow and water flow for the coils are the user-specified values instead of the autosized design values.

**Before:**
image

**After:**
image

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If structural output changes, add to output rules file and add OutputChange label

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally

### CODE REVIEW

File: src/EnergyPlus/WaterCoils.cc [hunk 1]
```diff
@@ -2209,9 +2209,20 @@ namespace WaterCoils {
...
+                        DataFlowUsedForSizing = ZoneEqSizing(CurZoneEqNum).AirVolFlow;
+                        WaterCoil(CoilNum).DesAirVolFlowRate = ZoneEqSizing(CurZoneEqNum).AirVolFlow;
+                    } else {
+                        DataAirFlowUsedForSizing = TempSize; // many autosized inputs use the design (autosized) air volume flow rate, save this value
+                        DataFlowUsedForSizing = TempSize;
+                    }
+                } else {
+                    DataAirFlowUsedForSizing = TempSize; // many autosized inputs use the design (autosized) air volume flow rate, save this value
+                    DataFlowUsedForSizing = TempSize;
+                }
```
Comment (rraustad) [inline src/EnergyPlus/WaterCoils.cc:2221]: This first change looks OK to me.

File: src/EnergyPlus/WaterCoils.cc [hunk 2]
```diff
@@ -2233,9 +2244,19 @@ namespace WaterCoils {
                 TempSize = AutoSize;
                 RequestSizing(state, CompType, CompName, CoolingCapacitySizing, SizingString, TempSize, bPRINT, RoutineName);
                 DataCapacityUsedForSizing = TempSize;
-                TempSize = WaterCoil(CoilNum).MaxWaterVolFlowRate;
+                TempSize = AutoSize;
```
Comment (rraustad) [inline src/EnergyPlus/WaterCoils.cc]: This second change always autosizes the HW coil water flow rate. What if the coil input is hard-sized? And the coil is on the main branch. I am on the fence whether this is OK for zone equipment, but if the fan coil HW flow rate is hard-sized or autosized then that flow rate will be used in the HW coil, even if the HW coil hard-sized the HW flow rate. This is an input over-specification problem and this does correct that. Do all other zone equipment also pass the ZoneEqSizing data, I assume they do, so again this is probably OK.

File: src/EnergyPlus/WaterCoils.cc [hunk 3]
```diff
@@ -2579,7 +2600,7 @@ namespace WaterCoils {
 
                 FieldNum = 2; // N2 , \field Maximum Water Flow Rate
                 SizingString = WaterCoilNumericFields(CoilNum).FieldNames(FieldNum) + " [m3/s]";
-                TempSize = WaterCoil(CoilNum).MaxWaterVolFlowRate;
+                TempSize = AutoSize;
```
Comment (rraustad) [inline src/EnergyPlus/WaterCoils.cc]: Same comment as above for always autosizing the HW flow rate whether or not the HW coil input is hard-sized.

**Comment by rraustad** [comment]
It is interesting that only the fan coil example files show diff's. However, there may not be example files where a water coil on the main branch is hard-sized. I think a quick test of 5ZoneAirCooled using hard-sized water flow rates for the main branch coils would show any problems. Also, what if a test file is completely hard-sized? And there are no sizing objects? That file would fail with a warning to add sizing objects. And users would say "huh".

**Comment by rraustad** [comment]
The fan coil was already passing the information. For some reason the water coil was not picking that information up in RequestSizing.

    ZoneEqSizing(CurZoneEqNum).MaxHWVolFlow = FanCoil(FanCoilNum).MaxHotWaterVolFlow;
    ZoneEqSizing(CurZoneEqNum).MaxCWVolFlow = FanCoil(FanCoilNum).MaxColdWaterVolFlow;

I think if you step through the sizing of a fan coil's water coil, by stepping into RequestSizing, you will see where the problem lies.

**Comment by matthew-larson** [comment]
Thanks @rraustad, I'll go back and step through the sizing portion and see if I can find the disconnect.

**Comment by nrel-bot-2** [comment]
@matthew-larson @lgentile it has been 28 days since this pull request was last updated.

**Comment by matthew-larson** [comment]
@rraustad I ended up removing the changes related to the water flow rates for this pull request.  I felt that I was getting beyond the scope of these issues which are addressing the design air flow rates and need to spend more time investigating RequestSizing for the water flow rates as you recommended previously.  Thanks again for your assistance with this.  Let me know if you have any additional comments regarding this pull request.

**Comment by rraustad** [comment]
Well, the logic here looks better. Good thing this is in water coils, because I have removed RequestSizing in the Autosizing Library branch. Can you paste in the eio diff's from a defect file so I have a reference for what you think is the correct result? And tell me what defect file or example file you used.

**Comment by matthew-larson** [comment]
I used the [_FanCoilHybridVentAFN.idf](https://github.com/NREL/EnergyPlus/files/5085876/_FanCoilHybridVentAFN.txt) testfile to show the effect.  The ZoneHVAC:FourPipeFanCoil object is autosized to a Design Size Maximum Supply Air Flow of 0.21702 m3/s.  The Coil:Cooling:Water associated with the fan coil specifies a Design Size Design Air Flow Rate of 0.21702 m3/s but a User-Specified Design Air Flow Rate of 0.12390 m3/s, which is used for the sizing of the coil, leaving the coil undersized compared to the fan coil specifications.  After the fix, the Coil:Cooling:Water gets a User-Specified Design Air Flow Rate of 0.21702 m3/s to match the size determined by the ZoneHVAC:FourPipeFanCoil object.  You can see the change in Nominal Total Capacity of the cooling coil and other design parameters below (before on left, after on right):

image

I also tested when the Maximum Supply Air Flow Rate for the ZoneHVAC:FourPipeFanCoil object is hard-sized to 0.2 m3/s instead of autosized.  Originally, hard-sizing this has no effect on the Coil:Cooling:Water Nominal Total Capacity as it still uses the User-Specified Design Air Flow Rate of 0.12390 m3/s.  After the fix, the Coil:Cooling:Water gets a User-Specified Design Air Flow Rate of 0.2 m3/s to match the size determined by the ZoneHVAC:FourPipeFanCoil object and increases the Nominal Total Capacity.  Same before and after comparison with the hard-sized flow rate shown below:

image

Let me know if you need any other info/results or have any additional or followup questions.

**Comment by rraustad** [comment]
Well, this goes to what the user intended. What if the user wanted to under (or over) size the cooling coil? With these changes there is no way to size a coil different from the fan coil size. Using the autosize method, all inputs should be autosized. If a user specifies a hard-size then E+ uses that information, regardless of result.

**Comment by matthew-larson** [comment]
I see what you mean, I need to address the ability to hard-size the water coil to be different than the fan coil.  I'll spend more time on this, thanks.

**Comment by matthew-larson** [comment]
I made a number of updates to get the eio results to reflect what I believe we want to see based on different scenarios of autosizing and hardsizing the fan coil and cooling coil inputs.  Below is a table showing the eio results for the ZoneHVAC:FourPipeFanCoil Maximum Supply Air Flow Rate and Coil:Cooling:Water Design Air Flow Rate using the same test file ([_FanCoilHybridVentAFN.idf](https://github.com/NREL/EnergyPlus/files/5085876/_FanCoilHybridVentAFN.txt)).

image

The warnings and failed regressions tests are all due to the Coil:Cooling:Water now reporting 'Design Size Design Air Flow Rate' instead of 'User-Specified Design Air Flow Rate' when the coil is connected to a fan coil which are both autosized for air flow rate, which I think is appropriate.  There is an error added to the results for '_FanCoilHybridVentAFN.idf' due to the change in the Coil:Cooling:Water air flow rate but wanted to get this all reviewed before addressing this if needed.

File: src/EnergyPlus/WaterCoils.cc [hunk 4]
```diff
@@ -5941,10 +5976,10 @@ namespace WaterCoils {
             UtilityRoutines::SameString(CoilType, "Coil:Cooling:Water")) {
             WhichCoil = UtilityRoutines::FindItem(CoilName, WaterCoil);
             if (WhichCoil != 0) {
-                if (UtilityRoutines::SameString(CoilType, "Coil:Cooling:Water") && WaterCoil(WhichCoil).DesAirVolFlowRate < 0.0) {
+                if (WaterCoil(WhichCoil).DesAirVolFlowRate <= 0.0) {
                     WaterCoil(WhichCoil).DesAirVolFlowRate = CoilDesFlow;
                 } else {
-                    WaterCoil(WhichCoil).DesAirVolFlowRate = CoilDesFlow;
+                    //WaterCoil(WhichCoil).DesAirVolFlowRate = CoilDesFlow;
```
Comment (Myoldmopar) [inline src/EnergyPlus/WaterCoils.cc:5982]: I guess the design air vol flow rate will just stay at zero for this case?

**Review by Myoldmopar — APPROVED** [review]

**Comment by Myoldmopar** [comment]
I've got no problem with this.  I _think_ all the review comments have been addressed.  CI results look OK given the expected string change.  I built locally with develop and ran tests and everything is still fine.  I'm inclined to merge but will hold for a while to see if anyone has more thoughts.

**Comment by Myoldmopar** [comment]
Merging.  Thanks @matthew-larson for the fix and @rraustad for looking at the changes multiple times.
